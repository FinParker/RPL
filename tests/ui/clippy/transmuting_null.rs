//@revisions: inline normal
//@[inline] compile-flags: -Z mir-opt-level=0 -Z inline-mir=true
//@[inline] ignore-on-host
//@[normal] compile-flags: -Z mir-opt-level=0 -Z inline-mir=false
#![allow(dead_code)]
#![allow(clippy::zero_ptr)]
#![allow(clippy::transmute_ptr_to_ref)]
#![allow(clippy::eq_op, clippy::missing_transmute_annotations)]
#![allow(clippy::manual_dangling_ptr)]

// Easy to lint because these only span one line.
// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn one_liners() {
    unsafe {
        let _: &u64 = std::mem::transmute(0 as *const u64);
        //~^ transmuting_null

        let _: &u64 = std::mem::transmute(std::ptr::null::<u64>());
        //~^ transmuting_null
    }
}

pub const ZPTR: *const usize = 0 as *const _;
pub const NOT_ZPTR: *const usize = 1 as *const _;

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn transmute_const() {
    unsafe {
        // Should raise a lint.
        let _: &u64 = std::mem::transmute(ZPTR);
        //~^ transmuting_null

        // Should NOT raise a lint.
        let _: &u64 = std::mem::transmute(NOT_ZPTR);
    }
}

fn main() {
    one_liners();
    transmute_const();
}
