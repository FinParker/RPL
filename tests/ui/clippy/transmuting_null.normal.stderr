error: transmuting a known null pointer into a reference
  --> tests/ui/clippy/transmuting_null.rs:15:23
   |
LL |         let _: &u64 = std::mem::transmute(0 as *const u64);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `0 as *const u64` is known to be a null pointer, so transmuting it into a reference is undefined behavior
   = note: `#[deny(rpl::transmuting_null)]` on by default

error: transmuting a known null pointer into a reference
  --> tests/ui/clippy/transmuting_null.rs:18:23
   |
LL |         let _: &u64 = std::mem::transmute(std::ptr::null::<u64>());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `std::ptr::null::<u64>()` is known to be a null pointer, so transmuting it into a reference is undefined behavior

error: transmuting a known null pointer into a reference
  --> tests/ui/clippy/transmuting_null.rs:30:23
   |
LL |         let _: &u64 = std::mem::transmute(ZPTR);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `ZPTR` is known to be a null pointer, so transmuting it into a reference is undefined behavior

error: aborting due to 3 previous errors

